op_crates/fetch/21_blob.js

```diff
@@ -0,0 +1,289 @@
+      c = input.charAt(++position)
+    );
+    return { result: input.slice(start, position), position };
+  }
+
+  /**
+   * @param {string} s
+   * @returns {string}
+   */
+  function convertLineEndingsToNative(s) {
+    const nativeLineEnding = isWindows ? "\r\n" : "\n";
+
+    let { result, position } = collectCodepointsNotCRLF(s, 0);
+
+    while (position < s.length) {
+      const codePoint = s.charAt(position);
+      if (codePoint == "\r") {
+        result += nativeLineEnding;
+        position++;
+        if (position < s.length && s.charAt(position) == "\n") {
+          position++;
+        }
+      } else if (codePoint == "\n") {
+        position++;
+        result += nativeLineEnding;
+      }
+      const { result: token, position: newPosition } = collectCodepointsNotCRLF(
+        s,
+        position,
+      );
+      position = newPosition;
+      result += token;
+    }
+
+    return result;
+  }
+
+  /**
+   * @param  {...Uint8Array} bytesArrays
+   * @returns {Uint8Array} 
+   */
+  function concatUint8Arrays(...bytesArrays) {
+    let byteLength = 0;
+    for (const bytes of bytesArrays) {
+      byteLength += bytes.byteLength;
+    }
+    const finalBytes = new Uint8Array(byteLength);
+    let current = 0;
+    for (const bytes of bytesArrays) {
+      finalBytes.set(bytes, current);
+      current += bytes.byteLength;
+    }
+    return finalBytes;
+  }
+
+  const utf8Encoder = new TextEncoder();
+  const utf8Decoder = new TextDecoder("utf8");
+
+  /** @typedef {BufferSource | Blob | string} BlobPart */
+
+  /** 
+     * @param {(BlobPart)[]} parts
+     * @param {string} endings
+     * @returns {Uint8Array}
+     */
+  function processBlobParts(parts, endings) {
+    /** @type {Uint8Array[]} */
+    const bytesArrays = [];
+    for (const element of parts) {
+      if (element instanceof ArrayBuffer) {
+        bytesArrays.push(new Uint8Array(element.slice(0)));
+      } else if (ArrayBuffer.isView(element)) {
+        const buffer = element.buffer.slice(
+          element.byteOffset,
+          element.byteOffset + element.byteLength,
+        );
+        bytesArrays.push(new Uint8Array(buffer));
+      } else if (element instanceof Blob) {
+        bytesArrays.push(
+          new Uint8Array(element[_byteSequence].buffer.slice(0)),
+        );
+      } else if (typeof element == "string") {
+        let s = element;
+        if (endings == "native") {
+          s = convertLineEndingsToNative(s);
+        }
+        bytesArrays.push(utf8Encoder.encode(s));
+      } else {
+        throw new TypeError("Unreachable code (invalild element type)");
+      }
+    }
+    return concatUint8Arrays(...bytesArrays);
+  }
+
+  /**
+   * @param {string} str 
+   * @returns {string}
+   */
+  function normalizeType(str) {
+    let normalizedType = str;
+    if (!/^[\x20-\x7E]*$/.test(str)) {
+      normalizedType = "";
+    }
+    return normalizedType.toLowerCase();
+  }
+
+  const _byteSequence = Symbol("[[ByteSequence]]");
```

Comment:
I've been declaring the internal slots as a member of the class (with no init) so they can have a JSDoc type.

Reply:
I tried that, but then TSC screams at me for redeclaring a property for some reason.

Reply:
😕 hmmmm... didn't get that with streams

Reply:
Yeah I was suprised too. I can get you an exact error in a few minutes.

Reply:
```
Subsequent property declarations must have the same type.  Property '[_byteSequence]' must be of type 'any', but here has type 'Uint8Array'.ts(2717)

21_blob.js(188, 7): '[_byteSequence]' was also declared here.
```

Reply:
Well that clearly means they aren't being declared as the same type though, right?  You need to use `/* @type {Uint8Array} */` then, right?

Reply:
@kitsonk I urge you to try it. I declared `[_byteSequence];` with `/* @type {Uint8Array} */` and all other places where we assign to `this[_byteSequence]`, we only assign strictly typed `Uint8Array`, but the error won't go away.

Reply:
Ok, for some reason in vscode we weren't getting another diagnostic that is the "cause" of it.  It appears that assignment in the constructor needs an explicit type annotation, even if the type being assigned is ignored, otherwise it will imply `any`.  I am not sure why that is the case.  So that was conflicting.

Also, the type annotation in the constructor on bytes is unnecessary, so if you do this in the constructor it works and the diagnostic goes away:

```diff
-      /** @type {Uint8Array} */
       const bytes = processBlobParts(parts, endings);
+      /** @type {Uint8Array} */
       this[_byteSequence] = bytes;
```

Or you could do this:

```diff
-      /** @type {Uint8Array} */
-       const bytes = processBlobParts(parts, endings);
-       this[_byteSequence] = bytes;
+      /** @type {Uint8Array} */
+      this[_byteSequence] = processBlobParts(parts, endings);
```

Reply:
(didn't run into this in streams, because none of the slots get initialized in the constructor, they get set in sub functions and TypeScript doesn't freak out for some reason... 😒 )